lib3h/src/p2p/p2p_gateway.rs

```diff
@@ -0,0 +1,258 @@
+#![allow(non_snake_case)]
+
+use crate::{
+    dht::{
+        dht_event::{DhtEvent, PeerHoldRequestData},
+        dht_trait::Dht,
+        rrdht::RrDht,
+    },
+    p2p::p2p_protocol::P2pProtocol,
+    transport::{
+        error::TransportResult,
+        protocol::{TransportCommand, TransportEvent},
+        transport_trait::Transport,
+        TransportId, TransportIdRef,
+    },
+    transport_dna::TransportDna,
+    transport_wss::TransportWss,
+};
+use lib3h_protocol::{Address, DidWork, Lib3hResult};
+
+/// Gateway to a P2P network.
+/// Enables Connections to many other nodes.
+/// Tracks distributed data for that P2P network.
+pub struct P2pGateway {
+    transport: Box<Transport>,
+    dht: Box<Dht>,
```

Comment:
I vote for using Generics rather than Trait Objects for composition, so we don't have to deal with all the restrictions and limitations of Trait Objects.

Reply:
You mean like this?
```Rust
pub struct P2pGateway<T: Transport, D: Dht> {
    transport: T,
    dht: D,
}
```